h/h_api/bulk_api/command_batch.py

```diff
@@ -0,0 +1,88 @@
+"""Batching and dispatching jobs to work on."""
+
+from contextlib import contextmanager
+
+from h.h_api.exceptions import CommandSequenceException
+
+
+class CommandBatch:
+    """A batch of commands.
+
+    This object will group similar commands together and decide when they
+    should be executed. The number of commands should always be a number we
+    are comfortable dealing with in memory at once.
+    """
+
+    def __init__(self, on_flush, batch_size=100):
+        """
+        :param on_flush: Callback to call when it's time to execute
+        :param batch_size: The max size of a batch
+        """
+
+        self.on_flush = on_flush
+
+        self.batch = None
+        self.batch_size = batch_size
+        self.finished_tasks = set()
+        self.current_task = None
+
+    def flush(self):
+        """Flush the current batch of commands to the callback."""
+
+        if not self.batch:
+            self.batch = None
+            return
+
+        command_type, data_type = self.current_task
+        self.on_flush(command_type, data_type, self.batch)
+
+        self.batch = None
```

Comment:
Should `flush()` be public? Does user code ever call it, or is it only called by `add()` below?

Reply:
It's mentioned in the `add()` documentation, but it's the callers responsibility to call `flush()` when all is said and done to ensure the last batch is flushed out.

There's no way to know when we are called with `add()` than no further adds will happen in the future.